test-suite/testkit/tests/api/api_service.rs

```diff
@@ -0,0 +1,124 @@
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Service with API, but without any transactions.
+//! This service can be used for testing features related only to the API.
+
+use chrono::{TimeZone, Utc};
+use serde_derive::{Deserialize, Serialize};
+
+use exonum::{
+    api,
+    runtime::{
+        rust::{
+            api::{ServiceApiBuilder, ServiceApiState},
+            DefaultInstance, Service,
+        },
+        InstanceId,
+    },
+};
+use exonum_derive::*;
+
+pub const SERVICE_NAME: &str = "api-service";
+pub const SERVICE_ID: InstanceId = 3;
+
+#[exonum_interface]
+pub trait ApiServiceInterface {}
+
+impl ApiServiceInterface for ApiService {}
+
+/// Sample query supported by API.
+#[derive(Debug, Clone, Copy, PartialEq, Default)]
+#[derive(Serialize, Deserialize)]
+pub struct PingQuery {
+    pub value: u64,
+}
+
+#[derive(Debug, Clone, Copy)]
+struct Api;
+
+impl Api {
+    /// Returns the same number that was in query.
+    fn ping_pong(_state: &ServiceApiState<'_>, ping: PingQuery) -> api::Result<u64> {
+        Ok(ping.value)
+    }
+
+    /// Returns `Gone` error.
+    fn gone(_state: &ServiceApiState<'_>, _ping: PingQuery) -> api::Result<u64> {
+        Err(api::Error::Gone)
+    }
+
+    fn wire(builder: &mut ServiceApiBuilder) {
+        let public_scope = builder.public_scope();
+
+        // Normal endpoint.
+        public_scope.endpoint("ping-pong", Self::ping_pong);
+
+        // Deprecated endpoints.
+        public_scope
+            .deprecated_endpoint("ping-pong-deprecated", None, Self::ping_pong)
+            .deprecated_endpoint(
+                "ping-pong-deprecated-with-deadline",
+                Some(Utc.ymd(2055, 12, 31).and_hms(23, 59, 59)),
+                Self::ping_pong,
+            )
+            .deprecated_endpoint_mut("ping-pong-deprecated-mut", None, Self::ping_pong);
+
+        // Gone endpoints.
+        public_scope
+            .endpoint_mut("gone-mutable", Self::gone)
+            .endpoint("gone-immutable", Self::gone);
+
+        // Moved endpoints.
+        let url_base = public_scope.url_base();
+        let moved_mutable_new_location = format!("{}/{}", &url_base, "ping-pong-deprecated-mut");
+        let moved_immutable_new_location = format!("{}/{}", &url_base, "ping-pong");
+        public_scope
+            .endpoint_mut(
+                "moved-mutable",
+                move |_state: &ServiceApiState<'_>, _query: PingQuery| -> api::Result<u64> {
+                    Err(api::Error::MovedPermanently(
+                        moved_mutable_new_location.clone(),
```

Comment:
Needless `clone` ?

Reply:
It's required, since it's a `move` closure.